ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -225,33 +219,34 @@ private String formatQualifiedName(QualifiedName name) {
     }
 
     protected Pair<String, Schema> visitLongLiteral(LongLiteral node, Boolean unmangleNames) {
-      functionArguments.addArgumentType(Schema.Type.INT64);
       return new Pair<>("Long.parseLong(\"" + node.getValue() + "\")", Schema.INT64_SCHEMA);
     }
 
     @Override
     protected Pair<String, Schema> visitIntegerLiteral(final IntegerLiteral node,
                                                        final Boolean context) {
-      functionArguments.addArgumentType(Schema.Type.INT32);
       return new Pair<>("Integer.parseInt(\"" + node.getValue() + "\")", Schema.INT32_SCHEMA);
     }
 
     @Override
     protected Pair<String, Schema> visitFunctionCall(FunctionCall node, Boolean unmangleNames) {
-      String functionName = node.getName().getSuffix();
-      UdfFactory udfFactory = functionRegistry.getUdfFactory(functionName);
-      functionArguments.beginFunction();
+      final String functionName = node.getName().getSuffix();
+      final UdfFactory udfFactory = functionRegistry.getUdfFactory(functionName);
+      final List<Schema.Type> argumentTypes = new ArrayList<>();
+      final ExpressionTypeManager expressionTypeManager
+          = new ExpressionTypeManager(schema, functionRegistry);
 
       String instanceName = functionName + "_" + functionCounter++;
 
       final String arguments = node.getArguments().stream()
           .map(arg -> process(arg, unmangleNames).getLeft())
           .collect(Collectors.joining(", "));
-      final List<Schema.Type> types = functionArguments.endFunction();
-      final Schema returnType = udfFactory.getFunction(types).getReturnType();
+      node.getArguments().forEach(arg -> argumentTypes.add(
+          expressionTypeManager.getExpressionType(arg).type()
```

Comment:
Looks like every use of `getExpressionType` is followed by a `type()` call. May be worth thinking about adding another appropriately named method to `ExpresssionTypeManager` which makes this `type` call directly and returns the value.

Reply:
+1

Reply:
I'll add a method for this.